### PR TITLE
os/bluestore: do not cache shard keys

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -4460,7 +4461,9 @@ int BlueStore::fsck(bool deep)
       }
       for (auto& s : o->extent_map.shards) {
 	dout(20) << __func__ << "    shard " << *s.shard_info << dendl;
-	expecting_shards.push_back(s.key);
+	string key;
+	get_extent_shard_key(o->key, s.offset, &key);
+	expecting_shards.push_back(key);
```
Comment (allensamuels) [inline src/os/bluestore/BlueStore.cc]: Yuck. How about:

expecting_shards.resize(expecting_shards.size()+1);
get_extent_shard_key(o->key,s.offset,&expecting_shards[expecting_shards.size()-1]);

This avoids the extra malloc/dealloc for the string on the stack.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -1629,7 +1629,7 @@ bool BlueStore::ExtentMap::update(KeyValueDB::Transaction t,
   } else {
 
     struct dirty_shard_t{
-      string* key = nullptr;
+      uint32_t offset;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Please add =0

**Comment by liewegas** [comment]
fixed.  also reordered Shard fields to pack properly.

**Comment by markhpc** [comment]
This is resulting in at least a 10% performance improvement for 4K random writes and significant memory savings when using smaller shard_max / target sizes.

**Comment by allensamuels** [comment]
Yes, the real problem was that the Shard contained a string which wasn’t placed into the mempool’s memory computations.


Allen Samuels
SanDisk |a Western Digital brand
2880 Junction Avenue, San Jose, CA 95134
T: +1 408 801 7030| M: +1 408 780 6416
allen.samuels@SanDisk.com<mailto:allen.samuels@SanDisk.com>

**Comment by ifed01** [comment]
@liewegas - forget to push?

**Comment by liewegas** [comment]
@ifed01 nope, latest is pushed..

**Comment by ifed01** [comment]
That's strange, can't see neither my nor Allen's comment resolved...

**Comment by liewegas** [comment]
oops, there it is

**Pull request merged**